### PR TITLE
Add a HistoricEncounterData control

### LINKED ISSUE #2025: JSONForms Control: Display values from previous encounters
Motivation: In vacination programs it is common that you have multiple dose of vaccines in different encounters. Similiar, for infants you have multiple follow up visits.
One solution to do this is to have special encounters for every visit, e.g. VaccineEncounter1, VaccineEncounter2, etc. However, this could become quite cumbersome. For example: if an encounter has been missed two vaccines might be given at the same time and this would require to create two encounter durring the same visit. Furthermore, you may want to see all vaccines that has been given so far, switching back and forth between encounter is not very user friendly.

Suggested solution:
- [ ] Have all encounter steps visible in a single encounter
- [ ] Display previous steps from previous encounters as readonly
- [ ] Use events to keep track which steps have been done, e.g. have one required field for the section and if it is truthly set an event that the section has been done. The use a special control to display the section as read only (see below).
- [ ] ~Bonus: Add option to redo a step, e.g. a clear button which then allows to enter data for an already taken step~

mock_staged_encounter

## Previous Data Control:
Requirements:
When a configurable event is active the control should display data from the event's encounter as read only. If event is not set the data is editable for the current encounter.
- [ ] show source encounter date?
- [ ] Custom layout, e.g. use default JsonSchema syntax to configure which fields are displayed.

Example config:
```json
   {
      "type": "PreviousEncounterData", // better name?
      "label": "Vaccination 1",
      "scope": "#/properties/extension/properties/vaccination1",
      "options": {
        "conditions": [
          { "field": "field1", "isTruthly": true }
        ],
        "elements": [
          {
            "type": "Control",
            "scope": "properties/field1",
          },
          ...
        ]
      }
    }
```

### PR DESCRIPTION
**Pull request by clemens-msupply**
Fixes #2025 

# 👩🏻‍💻 What does this PR do? 
Add a HistoricEncounterData control (better name?). Please see docs comment for the control, i.e. if the comments are clear.

image

Known issues:
- Left alignment for disabled Select component is wrong, any idea why?
- The datepicker can't be cleared, and in our example the date field is used to tricker the used program event, i.e. you can't undo a mistake and edit a section later... I tried to add the clear button to the date picker's action bar, but it didn't clear the underlying data... I think we should allow to clear it though, @CarlosNZ what do you think?

# 🧪 How has/should this change been tested? 
Use the matching branch from the schema repository and fill in some infant detection encounters, i.e. test that the "follow up" data entered previously can't be edited in a later encounter.

### CODE REVIEW

**Comment by CarlosNZ** [comment]
> * I tried to add the clear button to the date picker's action bar, but it didn't clear the underlying data... I think we should allow to clear it though, @CarlosNZ what do you think?

Does the `onChange` event not fire when clearing the date? If so, can we not just update the date as part of the `onChange` method?

Also, @clemens-msupply -- I can't see this component implemented anywhere in the programsschema repo -- should it be? How can we test it?

**Comment by clemens-msupply** [comment]
> > * I tried to add the clear button to the date picker's action bar, but it didn't clear the underlying data... I think we should allow to clear it though, @CarlosNZ what do you think?
> 
> Does the `onChange` event not fire when clearing the date? If so, can we not just update the date as part of the `onChange` method?

Yes right, just clearing it should be enough, will have another look...

> Also, @clemens-msupply -- I can't see this component implemented anywhere in the programsschema repo -- should it be? How can we test it?

There is a branch in the schema repo with the same name `2025-historic-encounter-data` with the updated schema

**Comment by clemens-msupply** [comment]
> > Left alignment for disabled Select component is wrong, any idea why?
> 
> Not sure what you mean - the text is left aligned. The left padding is missing when disabled but that's all I can see 🤷‍♂️

yes thats what I mean

File: client/packages/programs/src/JsonForms/components/HistoricEncounterData.tsx [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+            readonly={true}
+            renderers={renderers ?? []}
+          />
+          <FormLabel
+            sx={{
+              fontSize: '12px',
+              marginLeft: 'auto',
+            }}
+          >
+            {`Recorded on ${localisedDate(
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/HistoricEncounterData.tsx:131]: Thanks I change it and also translated the message

**Comment by CarlosNZ** [comment]
> Left alignment for disabled Select component is wrong, any idea why?

Yeah, that is odd. It's inline-ing `style="padding-left: 0px;"` to the  "Autocomplete" style when disabled.

I can override it by updating the "Autocomplete" style (in Select.tsx) to something like:
```
<Autocomplete
          sx={{
            '.MuiFormControl-root': { minWidth: '100%' },
            '.Mui-disabled.MuiInputBase-fullWidth': {
              paddingLeft: '10px !important',
            },
            flexBasis: '90%',
          }}
```

That can probably be improved somewhat, but it's a start.

File: client/packages/programs/src/JsonForms/components/HistoricEncounterData.tsx [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+          equalTo: patientId,
+        },
+        type: {
+          equalTo: options?.condition.eventType,
+        },
+      },
+      page: {
+        first: 1,
+      },
+    },
```
[the remark this answers is not in the document]
Comment (clemens-msupply) [reply client/packages/programs/src/JsonForms/components/HistoricEncounterData.tsx:83]: No it should't be needed there is only one event max for each type. Thanks.

**Comment by clemens-msupply** [comment]
Thanks @CarlosNZ actually found the reason:
image

@mark-prins do you remember the reason why this change was needed?

**Comment by mark-prins** [comment]
The change was in [this PR](https://github.com/openmsupply/openmsupply-client/pull/992)
Apparently, "the autocomplete used by the item selector wasn't left aligned for internal orders, so I tweaked."

I have an alternative though!

Remove the line from `Autocomplete`, and add this to the `Autocomplete` props in `StockItemSearchInputWithStats`

```
      sx={{
        '.MuiInputBase-root': { paddingLeft: disabled ? 0 : undefined },
        '.MuiBox-root': { justifyContent: 'flex-start' },
      }}
```

that looks better anyway - when enabled:

image

and disabled:

image

**Comment by clemens-msupply** [comment]
> The change was in [this PR](https://github.com/openmsupply/openmsupply-client/pull/992) Apparently, "the autocomplete used by the item selector wasn't left aligned for internal orders, so I tweaked."
> 
> I have an alternative though!
> 
> Remove the line from `Autocomplete`, and add this to the `Autocomplete` props in `StockItemSearchInputWithStats`
> 
> ```
>       sx={{
>         '.MuiInputBase-root': { paddingLeft: disabled ? 0 : undefined },
>         '.MuiBox-root': { justifyContent: 'flex-start' },
>       }}
> ```

Excellent! I added this. There is still some weird vertical spacing in the disabled form though, will have another look...

**Pull request merged**